### PR TITLE
[FEDEXSHIP-41] App Review

### PR DESCRIPTION
**Pull request by Arturo777**
- (FEDEXSHIP-41) Logging optimization
- (FEDEXSHIP-41) GraphQL mutation security
- (FEDEXSHIP-41) Admin navigation permissions

[Workspace to test](https://fedexlogger--sandboxusdev.myvtex.com/)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,8 +4,26 @@ All notable changes to this project will be documented in this file.
...
-
+## [Unreleased]
+
+### Added
+
+- (FEDEXSHIP-41) Logging optimization
+- (FEDEXSHIP-41) GraphQL mutation security
+- (FEDEXSHIP-41) Admin navigation permissions
+
+## [1.18.3] - 2022-09-06
```
Comment (polishq) [inline CHANGELOG.md:15]: Remove duplicate entry

File: dotnet/Controlers/RoutesController.cs [hunk 2]
```diff
@@ -45,66 +45,102 @@ public async Task<IActionResult> GetRates()
...
-
-            stopWatch.Stop();
-            TimeSpan ts = stopWatch.Elapsed;
-            _context.Vtex.Logger.Info("GetRates", "GetRates Time", "Time Spent in MS",
-                new[]
-                {
-                    ( "timeLapsed", ts.TotalMilliseconds.ToString()),
-                }
-            );
+            if (!getRatesResponseWrapper.IsValidCountry) {
```
Comment (polishq) [inline dotnet/Controlers/RoutesController.cs:49]:
```suggestion
            if (!getRatesResponseWrapper.IsValidCountry)
            {
```

File: dotnet/Controlers/RoutesController.cs [hunk 3]
```diff
@@ -45,66 +45,102 @@ public async Task<IActionResult> GetRates()
...
+                Response.StatusCode = 404;
```
Comment (polishq) [inline dotnet/Controlers/RoutesController.cs:50]: Add a return to stop the following code from being processed

File: dotnet/Services/FedExAvailabilityRequest.cs [hunk 4]
```diff
@@ -3,32 +3,39 @@
...
-            string remoteAddress = string.Empty;
             serviceAvailabilityResponse reply = new serviceAvailabilityResponse();
 
             try
             {
+
+                this._merchantSettings = await _merchantSettingsRepository.GetMerchantSettings();
+
+                ServiceAvailabilityRequest request = CreateServiceAvailabilityRequest();
+                //ValidationAvailabilityAndCommitmentService service = new ValidationAvailabilityAndCommitmentService(); // Initialize the service
```
Comment (polishq) [inline dotnet/Services/FedExAvailabilityRequest.cs:35]: Remove commented code

File: dotnet/Services/FedExRateRequest.cs [hunk 5]
```diff
@@ -101,190 +101,207 @@ public FedExRateRequest(IVtexEnvironmentVariableProvider environmentVariableProv
...
 
-                                Dictionary<string, double> ratesRatio = CalculateRatesRatio(copyItemResult);
-                                foreach(RateReplyDetail detail in reply.RateReplyDetails)
+                                getRatesResponseWrapper.timeSpan = ts;
+                                RateReply reply = ratesResponse.RateReply;
+                                cell.HighestSeverity.Add(reply.HighestSeverity.ToString());
+                                if (reply.HighestSeverity == NotificationSeverityType.SUCCESS || reply.HighestSeverity == NotificationSeverityType.NOTE || reply.HighestSeverity == NotificationSeverityType.WARNING)
                                 {
-                                    if (!slaMapping[detail.ServiceDescription.Description].Hidden)
+                                    // if (!string.Equals(this._environmentVariableProvider.Workspace, "master")) {
```
Comment (polishq) [inline dotnet/Services/FedExRateRequest.cs:211]: Remove commented code

File: dotnet/Services/FedExRateRequest.cs [hunk 6]
```diff
@@ -101,190 +101,207 @@ public FedExRateRequest(IVtexEnvironmentVariableProvider environmentVariableProv
...
+                                {
+                                    getRatesResponseWrapper.Notifications.Add(notification);
+                                }
                             }
-
-                            foreach(Notification notification in reply.Notifications)
+                            catch (Exception e)
                             {
-                                getRatesResponseWrapper.Notifications.Add(notification);
+                                Console.WriteLine($"Exception: {e.Message}");
```
Comment (polishq) [inline dotnet/Services/FedExRateRequest.cs:262]: Log to Splunk?

File: dotnet/Services/FedExRateRequest.cs [hunk 7]
```diff
@@ -101,190 +101,207 @@ public FedExRateRequest(IVtexEnvironmentVariableProvider environmentVariableProv
...
+                                cell.HighestSeverity.Add(reply.HighestSeverity.ToString());
+                                if (reply.HighestSeverity == NotificationSeverityType.SUCCESS || reply.HighestSeverity == NotificationSeverityType.NOTE || reply.HighestSeverity == NotificationSeverityType.WARNING)
                                 {
-                                    if (!slaMapping[detail.ServiceDescription.Description].Hidden)
+                                    // if (!string.Equals(this._environmentVariableProvider.Workspace, "master")) {
+                                    //     ShowRateReply(reply);
+                                    // }
+
+                                    Dictionary<string, double> ratesRatio = CalculateRatesRatio(copyItemResult);
+                                    foreach(RateReplyDetail detail in reply.RateReplyDetails)
```
Comment (polishq) [inline dotnet/Services/FedExRateRequest.cs:216]: Check for `null` `reply.RateReplyDetails` before performing `foreach`

File: dotnet/Services/FedExRateRequest.cs [hunk 8]
```diff
@@ -101,190 +101,207 @@ public FedExRateRequest(IVtexEnvironmentVariableProvider environmentVariableProv
...
                                     {
-                                        TimeSpan transitArrival = detail.DeliveryTimestamp - getRatesRequest.shippingDateUTC;
-                                        string transitString = new TimeSpan(transitArrival.Days, transitArrival.Hours, transitArrival.Minutes, transitArrival.Seconds).ToString();
-                                        foreach (Item item in copyItemResult)
+                                        if (!slaMapping[detail.ServiceDescription.Description].Hidden)
                                         {
-                                            GetRatesResponse rateResponse = new GetRatesResponse
+                                            TimeSpan transitArrival = detail.DeliveryTimestamp - getRatesRequest.shippingDateUTC;
```
Comment (polishq) [inline dotnet/Services/FedExRateRequest.cs:220]: This causes an error later if there is no `DeliveryTimestamp` in the response. You can check for a `detail.DeliveryTimestampSpecified` boolean in the response; if it's false, skip the SLA

**Review by polishq — CHANGES_REQUESTED** [review]

File: dotnet/Services/FedExRateRequest.cs [hunk 9]
```diff
@@ -101,190 +101,207 @@ public FedExRateRequest(IVtexEnvironmentVariableProvider environmentVariableProv
...
+                                    new[]
+                                    {
+                                        ( "timeLapsed", ts.TotalMilliseconds.ToString()),
+                                    }
+                                );
 
-                                Dictionary<string, double> ratesRatio = CalculateRatesRatio(copyItemResult);
-                                foreach(RateReplyDetail detail in reply.RateReplyDetails)
+                                getRatesResponseWrapper.timeSpan = ts;
+                                RateReply reply = ratesResponse.RateReply;
```
Comment (polishq) [inline dotnet/Services/FedExRateRequest.cs:207]: Let's log this `reply` please

**Review by polishq — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (Arturo777) [reply CHANGELOG.md:15]: done

File: dotnet/Services/FedExRateRequest.cs (hunk 5 above)
Comment (Arturo777) [reply dotnet/Services/FedExRateRequest.cs:211]: done

File: dotnet/Services/FedExRateRequest.cs (hunk 6 above)
Comment (Arturo777) [reply dotnet/Services/FedExRateRequest.cs:262]: done

File: dotnet/Services/FedExRateRequest.cs (hunk 8 above)
Comment (Arturo777) [reply dotnet/Services/FedExRateRequest.cs:220]: done

**Review by btalma — DISMISSED** [review]

**Review by polishq — APPROVED** [review]
LGTM

File: dotnet/Controlers/RoutesController.cs [hunk 10]
```diff
@@ -45,66 +42,105 @@ public async Task<IActionResult> GetRates()
...
-                "Get Rates Response", 
-                new[]
-                {
-                    ( "HighestSeverity", string.Join(",", getRatesResponseWrapper.HighestSeverity)),
-                    ( "Success", getRatesResponseWrapper.Success.ToString()),
-                    ( "Error", string.Join(",", getRatesResponseWrapper.Error)),
-                    ( "entireObject", JsonConvert.SerializeObject(getRatesResponseWrapper)),
-                }
-            );
+            if (!getRatesResponseWrapper.IsValidCountry)
```
Comment (polishq) [inline dotnet/Controlers/RoutesController.cs:46]: Ah, this is why the Cypress test is failing. You're checking for `IsValidCountry` before any API requests have been made. This needs to happen after `getRatesResponseWrapper = await this._fedExRateRequest.GetRates(getRatesRequest);`

**Review by polishq — CHANGES_REQUESTED** [review]

File: dotnet/Controlers/RoutesController.cs (hunk 10 above)
Comment (Arturo777) [reply dotnet/Controlers/RoutesController.cs]: ohhh, ok, i updated the PR!

**Review by polishq — APPROVED** [review]
LGTM

**Pull request merged**